### PR TITLE
Copy EC Publisher Proxy Code and Use  SSA

### PR DESCRIPTION
**Pull request by muralov**
**Description**
Copy EC publisher proxy code to Eventing Manager
Changes proposed in this pull request:

- Copy logic
- Copy tests
- Used Server-Side apply to simplify the code

**Related issue(s)**
#25

### CODE REVIEW

File: pkg/eventing/deployment_test.go [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+	// check eventmesh-specific env variables
+	eventMeshNamespace := findEnvVar(container.Env, "BEB_NAMESPACE")
+	assert.Equal(t, eventMeshNamespace.Value, fmt.Sprintf("%s$(BEB_NAMESPACE_VALUE)", eventMeshNamespacePrefix))
+
+	// check the affinity is empty
+	assert.Empty(t, deployment.Spec.Template.Spec.Affinity)
+}
+
+// findPublisherContainer gets the publisher proxy container by its name.
+func findPublisherContainer(publisherName string, deployment appsv1.Deployment) v1.Container {
```
Comment (mfaizanse) [inline pkg/eventing/deployment_test.go:293]: (nit) We can move it to test utils pkg so later other tests can use it as well?

File: pkg/eventing/deployment_test.go [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+		if strings.EqualFold(c.Name, publisherName) {
+			container = c
+		}
+	}
+	return container
+}
+
+// findEnvVar returns the env variable which has `name == envVar.Name`,
+// or `nil` if there is no such env variable.
+func findEnvVar(envVars []v1.EnvVar, name string) *v1.EnvVar {
```
Comment (mfaizanse) [inline pkg/eventing/deployment_test.go:305]: (nit) We can move it to test utils pkg so later other tests can use it as well?

File: pkg/eventing/manager.go [hunk 3]
```diff
@@ -50,111 +54,93 @@ func NewEventingManager(
...
-
-type ECReconcilerEventingClient struct {
-	ecReconciler *ecbackend.Reconciler
+func (em EventingManager) DeployPublisherProxy(ctx context.Context, eventing *v1alpha1.Eventing, backendType v1alpha1.BackendType) (*appsv1.Deployment, error) {
+	// update EC reconciler NATS and public config from the data in the eventing CR
+	if err := em.updateNatsConfig(ctx, eventing); err != nil {
+		return nil, err
+	}
+	em.updatePublisherConfig(eventing)
+	// deployment, err := em.ecReconcilerClient.DeployPublisherProxy(ctx, ecBackendType, em.natsConfig, em.backendConfig)
```
Comment (mfaizanse) [inline pkg/eventing/manager.go:74]: remove commented line

File: pkg/eventing/manager.go [hunk 4]
```diff
@@ -50,111 +54,93 @@ func NewEventingManager(
...
-	// update EC reconciler NATS and public config from the data in the eventing CR
-	if err := em.updateNatsConfig(ctx, eventing); err != nil {
-		return nil, err
+	if currentPublisher != nil {
+		// preserve only allowed annotations
+		desiredPublisher.Spec.Template.ObjectMeta.Annotations = make(map[string]string)
+		for k, v := range currentPublisher.Spec.Template.ObjectMeta.Annotations {
+			if _, ok := allowedAnnotations[k]; ok {
+				desiredPublisher.Spec.Template.ObjectMeta.Annotations[k] = v
+			}
```
Comment (mfaizanse) [inline pkg/eventing/manager.go:112]: Whats the purpose of this check? we dont allow users to add any annotations?

File: pkg/eventing/manager.go [hunk 5]
```diff
@@ -50,111 +54,93 @@ func NewEventingManager(
...
+	if err := em.kubeClient.PatchApply(ctx, desiredPublisher); err != nil {
+		return nil, fmt.Errorf("failed to apply Publisher Proxy deployment: %v", err)
 	}
-	return deployment, nil
+
+	return desiredPublisher, nil
+}
+
+// used for unit testing to mock the controllerutil.SetControllerReference
+var setOwnerReference = func(
```
Comment (mfaizanse) [inline pkg/eventing/manager.go:124]: Can we define it in the tests file where its used?

File: pkg/k8s/client.go [hunk 6]
```diff
@@ -37,13 +38,25 @@ func NewKubeClient(client client.Client, clientset k8sclientset.Interface, field
...
-	err := c.client.Get(ctx, client.ObjectKey{Name: name, Namespace: namespace}, deployment)
-	if err != nil {
-		return nil, err
+	if err := c.client.Get(ctx, client.ObjectKey{Name: name, Namespace: namespace}, deployment); err != nil {
+		return nil, client.IgnoreNotFound(err)
 	}
 	return deployment, nil
 }
 
+func (c *KubeClient) DeleteDeployment(ctx context.Context, name, namespace string) error {
```
Comment (mfaizanse) [inline pkg/k8s/client.go:47]: unit tests?

File: pkg/eventing/manager.go (hunk 4 above)
Comment (muralov) [reply pkg/eventing/manager.go:112]: This removes the not allowed annotations. Only `allowedAnnotations` are allowed to be set. I didn't come up with this idea, the logic comes from the EC.

File: pkg/eventing/manager.go (hunk 5 above)
Comment (muralov) [reply pkg/eventing/manager.go:124]: actually, it is used here not only in the tests. It is faked in the tests, but used in the line 97.

File: pkg/eventing/deployment_test.go (hunk 2 above)
Comment (muralov) [reply pkg/eventing/deployment_test.go:305]: I am already moving it for eventTypePrefix implementation, hence it'll be in another PR.

File: pkg/eventing/deployment_test.go (hunk 1 above)
Comment (muralov) [reply pkg/eventing/deployment_test.go:293]: sure, I'll move this with eventTypePrefix.

File: pkg/k8s/client.go (hunk 6 above)
Comment (muralov) [reply pkg/k8s/client.go:41]: done

**Review by mfaizanse — APPROVED** [review]

**Pull request merged**